### DISCUSSION TITLE
Bug: internal server error if somehow a route or dependency has 2 Request/Response/... dependencies

### DISCUSSION THREAD

**Discussion by volfpeter** [fastapi/fastapi#12403] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import Depends, FastAPI, Request, Response


def random_number(request: Request) -> int:
    return 4


app = FastAPI()


@app.get("/request")
def request_bug(r1: Request, r2: Request) -> str:
    # Internal server error, only 1 request argument will be passed.
    return "success"


@app.get("/response")
def response_bug(r1: Response, r2: Response) -> str:
    # Internal server error, only 1 response argument will be passed.
    return "success"


@app.get("/other")
def no_bug(r1: int = Depends(random_number), r2: int = Depends(random_number)) -> int:
    # No error, even though the same dependency is used twice here as well.
    return r1 + r2
```

### Description

If a route (or dependency) somehow has more than one "special" dependency of the same type (`Request`, `Websocket`, `HTTPConnection`, `Response`, `StarletteBackgroundTasks`, or `SecurityScopes`), then the route will fail with internal server error, because the current `Dependant` class only supports one argument name for these dependencies.

Note that this limitation does not exist for other dependencies.

I already have the fix, I just need to add some tests before creating a pull request.

### Operating System

Linux

### FastAPI Version

0.115.0

### Pydantic Version

2.9.2

### Python Version

3.12.5

### Additional Context

I understand that this seems to be a niche problem, but a user of one of my Python libraries reported a bug that occurred because of this, see his [PR here](https://github.com/volfpeter/fasthx/pull/34).

I know route decorators are not officially supported by FastAPI, but more and more packages pop up that implement their functionality with a decorator. Also, as mentioned in the description, there's no such limitation for "custom" dependencies (as can be seen in the example code). The bugfix is also very straightforward, it should have no negative effect on FastAPI.

I should also mention there are workarounds for this bug. For example creating a custom dependency that simply returns the `Request`/... object, and using this dependency with an `Any` or `object` annotation in routes and other dependencies. This is pretty ugly though.

**Comment by volfpeter** [comment, marked as answer]
For reference, here's the PR with the fix (and tests): https://github.com/fastapi/fastapi/pull/12406

**Comment by yakubka** [comment]
FastAPI treats `Request` and `Response` as special injection targets that do not go through Pydantic validation. When the same handler or dependency chain declares more than one `Request` parameter, FastAPI injects the same request object into both, which is fine at runtime but the dependency resolution code has a path where it fails to deduplicate the injection and raises an internal error.

The practical fix is to declare `Request` exactly once per handler.

```python
# wrong: two Request params
async def handler(req1: Request, req2: Request):
    ...

# correct: inject once, use everywhere
async def handler(request: Request):
    ...
```

If you need the request in a dependency and also in the handler, declare it in both function signatures independently. FastAPI will inject the same request object into each, and the error only occurs when the deduplication logic in the dependency resolver encounters an unexpected state. Using a single `Request` parameter per callable avoids that state entirely.

**Reply by volfpeter** [reply]
This is not always possible, for example in path operation decorators that need the request, see the linked PR/issue/fix for an example: https://github.com/volfpeter/fasthx/pull/34

FastHX already implements a clean workaround, but the bug should still be fixed in FastAPI. FWIW: the fix has been approved (for quite some time), hopefully it'll be merged this year...
